### PR TITLE
add explain and explain_instance methods to imputer

### PR DESCRIPTION
**Pull request by TammoR**
The imputer now exposes a method: .explain(label) that returns the top k tokens that are correlated with label, as well as .explain_instance(instance) that return the top k features that classify instance.

The necessary correlation matrices (training-samples x labels-classes) are computed during training if any explainable encoder is present. This also works when saving and then reloading a model.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/test_explain.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    imputer.predict(te)
+
+    # Evaluate
+    assert precision_score(te.out_cat, te.out_cat_imputed, average='weighted') > .99
+
+    # assert item explanation
+    text_explain = imputer.explain_instance(df.iloc[1])['in_text']
+    cat_explain = imputer.explain_instance(df.iloc[1])['in_cat']
+
+    if text_explain['top_class'] == 'bar':
```
Comment (dustin-lange) [inline test/test_explain.py]: Won't this test be deterministic - it's either always bar or always foo?

File: datawig/imputer.py [hunk 2]
```diff
@@ -244,8 +252,169 @@ def fit(self,
...
+        else:
+            label_encoder = self.label_encoders[0]
+
+        # Check whether to-be-explained label value exists.
+        if label not in label_encoder.token_to_idx.keys():
+            raise ValueError("Specified label {} not observed in label encoder".format(label))
+
+        # If k not given return all tokens
+        if k is None:
+            k = int(1e3)
```
Comment (andrey-tpt) [inline datawig/imputer.py]: Could you please add a comment that explains these 2 lines? Why 1000 ?

File: datawig/imputer.py [hunk 3]
```diff
@@ -244,8 +252,169 @@ def fit(self,
...
+        # assign index of label value (there can be an additional label column for "unobserved" label.
+        label_idx = label_encoder.token_to_idx[label]
+
+        # for each data encoder extract (token_idx, token_idx_correlation_with_label), extract and apply idx2token map.
+        token_weights = {}
+        for encoder, pattern in self.__class_patterns:
+            # extract idx2token mappings
+            if isinstance(encoder, CategoricalEncoder):
+                idx_tuples = zip(pattern[:, label_idx].argsort()[::-1][:k], sorted(pattern[:, label_idx])[::-1][:k])
+                # +1, bc idx_to_token has 0th column for unobserved labels. TODO: what does it do? remove this column!
```
Comment (andrey-tpt) [inline datawig/imputer.py]: Is "TODO" still necessary ?

File: datawig/imputer.py [hunk 4]
```diff
@@ -244,8 +252,169 @@ def fit(self,
...
+        Each entry shows the most likely class label for the input and the features with the highest correlation
+        with that class.
+
+        :param instance: row of data frame (or dictionary)
+        :param k: number of explanations (ngrams) for text inputs
+        :param label_column: name of label column to be explained (optional)
+        """
+
+        # assign label column encoder
+        if label_column is not None:
```
Comment (andrey-tpt) [inline datawig/imputer.py]: This looks suspiciously similar to beginning of _explain()_ method.  Could you extract to a method ?

File: test/test_explain.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    elif cat_explain['top_class'] == 'foo':
+        assert cat_explain['token_weights'][0][0] == 'foo'
+
+
+    # assert class explanations
+    assert np.all(['f' in token for token, weight in imputer.explain('foo')['in_text']][:3])
+    assert ['f' in token for token, weight in imputer.explain('foo')['in_cat']][0]
+
+    # test serialisation
+    imputer.save()
```
Comment (andrey-tpt) [inline test/test_explain.py]: What will happen with save() after the test ? If something already exists  in the location? There's test_dir feature which creates clean temp directory for test and then deletes it after the test

File: datawig/imputer.py [hunk 6]
```diff
@@ -28,12 +28,14 @@
 
 import mxnet as mx
 import numpy as np
+from sklearn.preprocessing import StandardScaler
```
Comment (andrey-tpt) [inline datawig/imputer.py]: I think this import statement should be on line number 35

File: datawig/imputer.py [hunk 7]
```diff
@@ -85,6 +87,12 @@ def __init__(self,
         self.precision_recall_curves = {}
         self.calibration_info = {}
 
+        self.__class_patterns = None
+        self.is_explainable = np.any([isinstance(encoder, CategoricalEncoder) or isinstance(encoder, TfIdfEncoder)
```
[the remark this answers is not in the document]
Comment (TammoR) [reply datawig/imputer.py]: I don't think there are obvious extensions to other input encoders, maybe to numerical. I don't think we will do this anytime soon, since it's not something we see in applications a lot.

File: datawig/imputer.py [hunk 8]
```diff
@@ -244,8 +252,169 @@ def fit(self,
...
+        p = self.__predict_mxnet_iter(iter_train)[label_name]  # class probabilities for every item (items x labels)
+
+        # center and whiten the class probabilities
+        p_normalized = StandardScaler().fit_transform(p)
+
+        # Generate list of data encoders, with features suitable for explanation. Only TfIDf and Categorical supported.
+        explainable_data_encoders = []
+        explainable_data_encoders_idx = []
+        for encoder_idx, encoder in enumerate(self.data_encoders):
+            if not (isinstance(encoder, TfIdfEncoder) or isinstance(encoder, CategoricalEncoder)):
```
[the remark this answers is not in the document]
Comment (TammoR) [reply datawig/imputer.py]: In is_explainable we only check whether any appropriate column exists. Here we need to select all these columns to compute their class-feature-correlations.

File: datawig/imputer.py [hunk 9]
```diff
@@ -244,8 +252,169 @@ def fit(self,
...
+                class_patterns.append((encoder, np.array(
+                        [np.sum(p_normalized[np.where(feature_matrix_scaled[0, :] == category)[0], :], axis=0)
+                         for category in encoder.idx_to_token.keys()])))
+            else:
+                logger.warn("column encoder not supported for explain.")
+
+        self.__class_patterns = class_patterns
+
+
+    def explain(self, label: str, k: int = None, label_column: str = None):
```
[the remark this answers is not in the document]
Comment (TammoR) [reply datawig/imputer.py]: good catch. 10 is more reasonable.

File: datawig/imputer.py [hunk 10]
```diff
@@ -244,8 +252,169 @@ def fit(self,
...
+        """
+        Return dictionary with entries for each explainable input column,
+        returning top k tokens with highest correlation to label.
+
+        :param label: label value to explain
+        :param k: number of explanations for each input encoder to return
+        :param label_column: name of label column to be explained (optional)
```
[the remark this answers is not in the document]
Comment (TammoR) [reply datawig/imputer.py]: Default picks the first available column. Will clarify in the docstring.

File: test/test_explain.py (hunk 1 above)
Comment (TammoR) [reply test/test_explain.py]: For a fixed random seed, yes. Should iterate over a few random rows instead.

File: datawig/imputer.py (hunk 4 above)
Comment (TammoR) [reply datawig/imputer.py]: good catch.

File: datawig/imputer.py [hunk 11]
```diff
@@ -244,8 +253,166 @@ def fit(self,
...
 
+        if self.is_explainable:
+            self.__persist_class_prototypes(iter_train, train_df)
+
         return self
 
+
+    def __persist_class_prototypes(self, iter_train, train_df):
+        """
+        Save mean feature pattern as self.class_patterns for each label_encoder, for each label, for each data encoder,
```
Comment (tdhd) [inline datawig/imputer.py]:
```suggestion
        Save mean feature pattern as self.__class_patterns for each label_encoder, for each label, for each data encoder,
```

File: datawig/imputer.py [hunk 12]
```diff
@@ -244,8 +253,166 @@ def fit(self,
...
+        given by the projection of the feature matrix (items by ngrams/categories)
+        onto the softmax outputs (items by labels).
+        self.class_patterns is a list of tuples of the form (column_encoder, feature-label-correlation-matrix).
```
Comment (tdhd) [inline datawig/imputer.py]:
```suggestion
        self.__class_patterns is a list of tuples of the form (column_encoder, feature-label-correlation-matrix).
```

File: datawig/imputer.py [hunk 13]
```diff
@@ -244,8 +253,166 @@ def fit(self,
...
+                idx_tuples = zip(pattern[:, label_idx].argsort()[::-1][:k], sorted(pattern[:, label_idx])[::-1][:k])
+                idx2token_temp = encoder.idx_to_token
+            token_weights[encoder.output_column] = [(idx2token_temp[token], weight) for token, weight in idx_tuples]
+
+        return token_weights
+
+
+    def explain_instance(self, instance: pd.core.series.Series, k: int = 10, label_column: str = None) -> dict:
+        """
+        Return dictionary with entries for each explainable input column of the give instance.
```
Comment (tdhd) [inline datawig/imputer.py]:
```suggestion
        Return dictionary with entries for each explainable input column of the given instance.
```
